packages/muelu/test/meshtying/MeshTyingBlocked_SimpleSmoother.cpp

```diff
@@ -0,0 +1,323 @@
+
+  Array<GO> myDualDofs(numMyDualDofs);
+  Array<GO> myDofs(numMyDofs);
+
+  for (size_t i = 0; i < numMyPrimalDofs; ++i)
+    myDofs[i] = myPrimalDofs[i];
+
+  size_t current_i = 0;
+  for (auto i = lagr2Dof.begin(); i != lagr2Dof.end(); ++i)
+    if (primalMap->isNodeGlobalElement(nPrimalDofsPerNode * (i->second)))
+    {
+      for (int j = 0; j < nDualDofsPerNode; ++j)
+      {
+        myDualDofs[current_i] = (i->first) * nDualDofsPerNode + j;
+        myDofs[numMyPrimalDofs + current_i] = globalPrimalNumDofs + (i->first) * nDualDofsPerNode + j;
+        ++current_i;
+      }
+      myLagr2Dof[i->first] = i->second;
+    }
+
+  RCP<const tpetra_map_type> dualMap = rcp(new tpetra_map_type(globalDualNumDofs, myDualDofs, indexBase, comm));
+  RCP<const tpetra_map_type> fullMap = rcp(new tpetra_map_type(globalNumDofs, myDofs, indexBase, comm));
+
+  RCP<const xpetra_map_type> fullXMap = rcp(new xpetra_tmap_type(fullMap));
+  RCP<const xpetra_map_type> primalXMap = rcp(new xpetra_tmap_type(primalMap));
+  RCP<const xpetra_map_type> dualXMap = rcp(new xpetra_tmap_type(dualMap));
+
+  std::vector<RCP<const xpetra_map_type>> xsubmaps = {primalXMap, dualXMap};
+  RCP<xpetra_bmap_type> blockedMap = rcp(new xpetra_bmap_type(fullXMap, xsubmaps, true));
+
+  // Read input matrices
+  typedef Tpetra::MatrixMarket::Reader<SparseMatrixType> reader_type;
+
+  RCP<xpetra_matrix_type> xQ = Xpetra::IO<Scalar, LocalOrdinal, GlobalOrdinal, Node>::Read("Q_mm.txt", primalXMap, null, primalXMap, primalXMap);
+  RCP<xpetra_matrix_type> xG = Xpetra::IO<Scalar, LocalOrdinal, GlobalOrdinal, Node>::Read("G_mm.txt", primalXMap, null, dualXMap, primalXMap);
+  RCP<xpetra_matrix_type> xGT = Xpetra::IO<Scalar, LocalOrdinal, GlobalOrdinal, Node>::Read("GT_mm.txt", dualXMap, null, primalXMap, dualXMap);
+  RCP<xpetra_matrix_type> xC = Xpetra::IO<Scalar, LocalOrdinal, GlobalOrdinal, Node>::Read("C_mm.txt", dualXMap, null, dualXMap, dualXMap);
+
+  RCP<xpetra_wcrs_type> xwQ = Teuchos::rcp_dynamic_cast<xpetra_wcrs_type>(xQ);
+  RCP<xpetra_wcrs_type> xwG = Teuchos::rcp_dynamic_cast<xpetra_wcrs_type>(xG);
+  RCP<xpetra_wcrs_type> xwGT = Teuchos::rcp_dynamic_cast<xpetra_wcrs_type>(xGT);
+  RCP<xpetra_wcrs_type> xwC = Teuchos::rcp_dynamic_cast<xpetra_wcrs_type>(xC);
+
+  // Construct the blocked saddle-point matrix
+  RCP<xpetra_bcrs_type> blockedMatrix = rcp(new xpetra_bcrs_type(blockedMap, blockedMap, 8));
+
+  blockedMatrix->setMatrix(0, 0, xwQ);
+  blockedMatrix->setMatrix(0, 1, xwG);
+  blockedMatrix->setMatrix(1, 0, xwGT);
+  blockedMatrix->setMatrix(1, 1, xwC);
+  blockedMatrix->fillComplete();
+
+  // Create the preconditioner
+  std::string xmlFile = "myXML.xml";
+
+  ParameterListInterpreter mueLuFactory(xmlFile, *comm);
+  RCP<Hierarchy> H = mueLuFactory.CreateHierarchy();
+  H->GetLevel(0)->Set("A", rcp_dynamic_cast<xpetra_matrix_type>(blockedMatrix));
+
+  H->GetLevel(0)->template Set<RCP<std::map<GO, GO>>>("Lagr2Dof", rcpFromRef(myLagr2Dof));
```

Comment:
This is not really the standard interface to pass parameters to MueLu, more of a hack-ish way to do so.
I would recommend actually dropping the map on the parameter list.

Reply:
I guess, the reason why @kliegeois did it that way is, that "Lagr2Dof" is information that needs to be provided on the finest level by the user (like "A"). One could probably put the dof map in the parameter list but probably not when using xml files (or json files), right? If you want to use the parameter list to provide the finest level information, you would need to read the xml file into a parameter list object and then add the "Lagr2Dof" map to the parameter list using C++ code. Or has this changed in between?

Reply:
@kliegeois I'm not quite sure whether "Lagr2Dof" is a very obvious name for the map you need. It is the mapping of the Lagrange node ids to dof ids, right? In between I'm tending to use longer but clearer variable names in order to produce "cleaner code" that is easier to understand and maintain for other people.

Reply:
@tawiesn both of your comments are right.
For the first one, as you mentioned, it would be difficult to write the map in xml files especially because this map is typically not known when writing the xml files and is an output of the computation of the matrices (with Moertel for instance).  However, it would definitely be possible to first read the xml files and add the map to the parameter list in C++.
For the second one, the map is the mapping of the Lagrange node ids to dof ids if there is one dof per mesh node and one Lagrange multipliers per interface node. Otherwise, it is the mapping of the Lagrange node ids to mesh node ids. 
If you prefer, we can use a map which would be the mapping between the Lagrange dof ids and the primal dof ids, however, this map would be larger.
I can definitely use a larger and clearer variable name here.

Reply:
@kliegeois Ok, i see. Storing the mapping of the Lagrange node ids to dof ids is fine. But you have those (reasonable) assumptions (e.g., one Lagrange multiplier per interface node). Some of them might not be immediately obvious for other people not so familiar with the code. Maybe you can add some comments about this in the code or description of the classes.

Reply:
@tawiesn there are three assumptions: the same number of dof per node (primal variable), the same number of dof per interface node (dual variable), and the fact that all dof related to a particular node (primal or dual) are in a same aggregate. The code, as it is, works when more than one Lagrange multiplier per interface node, for instance when solving contact problem with friction. I will clarify that in comments. However, it might be better to remove all those assumptions using a map which is the mapping between the dual dof ids and the primal dof ids. I do not have a strong feeling on the best option.

Reply:
@kliegois @tawiesn For the case of equal number of primal and dual DOFs (as for friction), I like the idea of just providing a node-to-node mapping. Aggregates are formed out of nodes, so nodes are the information that you want. Providing dog maps just to convert them manually back to node maps is not super straightforward in my opinion. 

A clarification in the documentation should suffice.

Reply:
Yes, from an aggregation perspective, using node maps makes more sense. Converting dof maps to node maps would just introduce additional overhead.